File: js/player.js
function Player(startEnergy, baseName, parent)
{
	this.scene = parent.scene;
	this.sceneMap = parent.sceneMap;

	this.camera = parent.camera;

    this.energy = startEnergy;
    this.energyGenerationSpeed = 0.2;

    this.baseName = baseName;

    this.selectedObject = null;
    this.selectedBase = null;

    this.bases = [];
    this.units = [];

	this.enemy = null;

    this.loader = new THREE.JSONLoader();

    this.addBase = function(unitSpawnPosition, mesh, color) {
        var base = new Base(unitSpawnPosition, mesh, this.scene, color);
        this.bases.push(base);
		this.selectedBase = base;
        return base;
    };

    this.addUnit = function(health, color) {
        this.units.push(new Unit0(health, color, this));
        this.energy -= this.selectedBase.unitCost;
    };

	this.removeUnit = function(unit) {
		var i = this.units.indexOf(unit);
		if (i === -1) {
			log("removeUnit index -1");
			return;
		}

		this.scene.remove(this.units[i].mesh);
		this.scene.remove(this.units[i].meshOutline);
		this.scene.remove(this.units[i].bullet.mesh);
		delete this.units[i].bullet;
		delete this.units[i];
		this.units.splice(i,1);
		this.selectedObject = null;
	};

    this.goUnit = function(point) {
        for ( var i = 0; i < this.units.length; i++) {
            if (this.units[i].mesh.position == this.selectedObject.position)
                this.units[i].goTo(new THREE.Vector3(point.x, this.units[i].mesh.position.y, point.z));
        }
    };

    this.deselectAll = function()
    {
        var i;
        for (i = 0; i < this.units.length; i++) this.units[i].select(false);
        for (i = 0; i < this.bases.length; i++) this.bases[i].select(false);
    };
    
    this.update = function(dt) {
        // generate energy
        this.energy += this.energyGenerationSpeed * dt;
        
        var i;
        for (i = 0; i < this.units.length; i++) {
			var unit = this.units[i];
			if (unit.health <= 1) this.removeUnit(unit); // 1 because I saw 0.25 health unit
			unit.update(dt);
		}
    };
}

File: JME/src/mygame/Player.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package mygame;

import com.jme3.ai.navmesh.NavMeshPathfinder;
import com.jme3.material.Material;
import com.jme3.math.ColorRGBA;
import com.jme3.math.Vector3f;
import com.jme3.scene.Geometry;
import com.jme3.scene.Node;
import com.jme3.scene.shape.Box;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 *
 * @author zDemoniac
 */
public class Player {
    private float energy = 0;
    private float energyGenerationSpeed = 0.2f;
    
    private ColorRGBA color;

    private String baseNamePrefix;

    private Node selectedObject = null;
    private Base selectedBase = null;
    private Game game = null;

    private Map<String, Base> bases = new HashMap<String, Base>();
    private List<Unit> units = new ArrayList<Unit>();

    
    public Player(ColorRGBA color, float startEnergy, String baseName, Game game) {
        energy = startEnergy;
        this.baseNamePrefix = baseName;
        this.game = game;
        this.color = color;
    }
    
    public void addBase(String name, Vector3f pos, int color) {
        Node baseNode = (Node)game.getRootNode().getChild(baseNamePrefix+".ogremesh");
        Base base = new Base(pos, baseNode);
        bases.put(name, base);
        
        // TODO
        selectedBase = base;
    }
    
    public void addUnit(String[] parts) {
        int price = 0;
    
        // check price, TODO: calc to GUI and get from there
        for (String part :parts) {
            //System.out.println(part);
            Base.PartInfo partInfo = selectedBase.getPartInfo(part);
            price += partInfo.getPrice();
        }

	if (energy >= price) {
            units.add(new Unit(parts, this));
	    energy -= price;
        }
    }
    
    public float getEnergy() {
        return energy;
    }
    
    public String getBaseNamePrefix() {
        return baseNamePrefix;
    }
    
    public Base getBase() {
        return selectedBase;
    }
    
    public Game getGame() {
        return game;
    }
    
    public ColorRGBA getColor() {
        return color;
    } 
    
    public Node getSelectedObject() {
        return selectedObject;
    }
    
    public void setSelectedObject(Node target) {
        selectedObject = target;
    }

    void update(float tpf) {
       energy += energyGenerationSpeed * tpf; 
       
       for(Unit unit: units) {
           unit.update(tpf);
       }
    }

    void goUnit(Vector3f pos) {
        Unit unit = selectedObject.getUserData("parent");
        unit.goTo(pos);
        }
}
